llvm/lib/Target/AIE/AIEBaseISelLowering.cpp

```diff
@@ -351,6 +365,23 @@ bool AIEBaseTargetLowering::isEligibleForTailCallOptimization(
   return true;
 }
 
+LLT AIEBaseTargetLowering::getOptimalMemOpLLT(
+    const MemOp &Op, const AttributeList &FuncAttributes) const {
+
+  if (Subtarget.isAIE2()) {
+    if (AlloVecRegMemOps && Op.size() >= 32 && Op.isAligned(Align(4)))
+      return LLT::fixed_vector(16, 16);
+    if (AlloVecRegMemOps && Op.size() >= 16 && Op.isAligned(Align(2)))
+      return LLT::fixed_vector(4, 32);
+    if (Op.size() >= 4 && Op.isAligned(Align(4)))
+      return LLT::scalar(32);
+    if (Op.size() >= 2 && Op.isAligned(Align(2)))
+      return LLT::scalar(16);
```

Comment:
I could do with explanation of the numbers and their units. In particular, I don't see how the alignment of a 16x16 vector could be the same as a scalar32

Reply:
> I could do with explanation of the numbers and their units. In particular, I don't see how the alignment of a 16x16 vector could be the same as a scalar32

Hi @martien-de-jong , correct observation. I was using the alignment information from  Legalizer:

```
  getActionDefinitionsBuilder({G_LOAD, G_STORE})
      .legalForTypesWithMemDesc({
          {S32, P0, S8, 8},         {S32, P0, S16, 16},
          {S20, P0, S20, 32},       {S32, P0, S32, 32},
          {P0, P0, S20, 32},        {V16S8, P0, V16S8, 16},
          {V8S16, P0, V8S16, 16},   {V4S32, P0, V4S32, 16},
          {V8S32, P0, V8S32, 32},   {V16S16, P0, V16S16, 32},
          {V32S8, P0, V32S8, 32},   {V16S32, P0, V16S32, 32},
          {V32S16, P0, V32S16, 32}, {V64S8, P0, V64S8, 32},
          {V32S32, P0, V32S32, 32}, {V64S16, P0, V64S16, 32},
          {V128S8, P0, V128S8, 32}, {ACC256, P0, ACC256, 32},
          {ACC512, P0, ACC512, 32}, {ACC1024, P0, ACC1024, 32},
          {S128, P0, S128, 16},
/*.....*/
```

Now, the test failures vanished...